src/target/at32f43x.c

```diff
@@ -0,0 +1,376 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2023 1BitSquared <info@1bitsquared.com>
+ * Written by ALTracer <tolstov_den@mail.ru>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file implements AT32F43x target specific functions for detecting
+ * the device, providing the XML memory map and Flash memory programming.
+ *
+ * References:
+ * ARTERY doc - RM_AT32F435_437_EN_V2.04.pdf
+ *   Reference manual - AT32F435/437 Series Reference Manual
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+#include "cortexm.h"
+
+static bool at32f43_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool at32f43_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+static bool at32f43_mass_erase(target_s *target);
+
+/* Flash memory controller register map */
+#define AT32F435_FLASH_REG_BASE 0x40023c00U
+#define AT32F435_FLASH_UNLOCK   (AT32F435_FLASH_REG_BASE + 0x04U)
+#define AT32F435_FLASH_STS      (AT32F435_FLASH_REG_BASE + 0x0cU)
+#define AT32F435_FLASH_CTRL     (AT32F435_FLASH_REG_BASE + 0x10U)
+#define AT32F435_FLASH_ADDR     (AT32F435_FLASH_REG_BASE + 0x14U)
+//#define AT32F435_FLASH_UNLOCK2  (AT32F435_FLASH_REG_BASE + 0x44U)
+//#define AT32F435_FLASH_STS2     (AT32F435_FLASH_REG_BASE + 0x4cU)
+//#define AT32F435_FLASH_CTRL2    (AT32F435_FLASH_REG_BASE + 0x50U)
+//#define AT32F435_FLASH_ADDR2    (AT32F435_FLASH_REG_BASE + 0x54U)
```

Comment:
As these are unused, please drop them rather than commenting them out. You'd be welcome to write a comment noting of their existence and being +0x40 from the previous registers, but this is a clang-tidy lint for a reason.

Reply:
you'll need to install clang-tidy and then you can get lints from the tool running `make clang-tidy` or, as that'll spew tons of output and be a bit unhelpful, if you use, say, VSCode then you can run this configuration to have the cpptools extension run clang-tidy for you per-file:
```json
	"C_Cpp.codeAnalysis.runAutomatically": true,
	"C_Cpp.codeAnalysis.clangTidy.enabled": true,
	"C_Cpp.codeAnalysis.clangTidy.useBuildPath": true,
```

You can technically also run clang-tidy manually per-file but getting the command line options right is a bit of a bore which is why there's automation in the repo for it (scripts/run-clang-tidy.py)

Reply:
I installed the pre-commit hook to run clang-format, can `clang-tidy` be run likewise for modified files or even diff context? `clang-format -i src/tu.c` updates sources per-file to conform to codestyle in its settings, and this is useful for me. I don't do codebase-wide refactors (often), and `make clang-tidy` outputs a ton of noise for me.
I run `cppcheck` as Geany lint button, so having `scripts/run-clang-tidy.py` print an example command would allow me to attach it into Geany User actions (it understands %f and other templates).

Reply:
Unsure what "%f" means in this context and especially as we don't know Geany, but we'd be surprised if it doesn't have some kind of clang-tidy integrations. Regarding pre-commit hooks, we'd have to ask @perigoso about those as we don't know but we also wouldn't include one in the repo for the same reason as why `make clang-tidy` is currently not so helpful.

We can look at having the run-clang-tidy.py script "dry run" things and look at specifying specific files to poke rather than the whole project.

Reply:
Our pre-commit configuration doesn't include a hook for `clang-tidy` at the moment, only `clang-format`, but that can be done and I'm pretty sure there's already one ready to go for us.
I don't remember why I didn't include it from the start, but if I had to guess it was because the code still contained a lot of `clang-tidy` warnings and it would probably cause more trouble than help